waspc/src/Wasp/Generator/DbGenerator.hs

```diff
@@ -99,7 +110,49 @@ genPrismaSchema wasp = createTemplateFileDraft dstPath tmplSrcPath (Just templat
 
 genMigrationsDir :: Wasp -> Maybe FileDraft
 genMigrationsDir wasp =
-  (getMigrationsDir wasp) >>= \waspMigrationsDir ->
+  getMigrationsDir wasp >>= \waspMigrationsDir ->
     Just $ createCopyDirFileDraft (SP.castDir genProjectMigrationsDir) (SP.castDir waspMigrationsDir)
   where
     genProjectMigrationsDir = dbRootDirInProjectRootDir </> dbMigrationsDirInDbRootDir
+
+writeDbSchemaChecksumToFile :: Path' Abs (Dir ProjectRootDir) -> IO ()
+writeDbSchemaChecksumToFile genProjectRootDir = do
+  dbSchemaExists <- doesFileExist dbSchemaFp
+  when dbSchemaExists $ do
+    checksum <- calculateChecksumStringForFile dbSchemaFp
+    writeFile dbSchemaChecksumFp checksum
+  where
+    dbSchemaFp = SP.fromAbsFile $ genProjectRootDir </> dbSchemaFileInProjectRootDir
+    dbSchemaChecksumFp = SP.fromAbsFile $ genProjectRootDir </> dbSchemaChecksumFileInProjectRootDir
+
+checkIfDbSchemaMatchesChecksum :: Path' Abs (Dir ProjectRootDir) -> ExceptT String IO ()
+checkIfDbSchemaMatchesChecksum genProjectRootDir = do
+  dbSchemaExists <- liftIO $ doesFileExist dbSchemaFp
+  let maybeDbSchemaFp = if dbSchemaExists then Just dbSchemaFp else Nothing
+  dbSchemaChecksumExists <- liftIO $ doesFileExist dbSchemaChecksumFp
+  let maybeDbSchemaChecksumFp = if dbSchemaChecksumExists then Just dbSchemaChecksumFp else Nothing
+  checkIfDbSchemaMatchesChecksum' maybeDbSchemaFp maybeDbSchemaChecksumFp
+  where
+    dbSchemaFp = SP.fromAbsFile $ genProjectRootDir </> dbSchemaFileInProjectRootDir
+    dbSchemaChecksumFp = SP.fromAbsFile $ genProjectRootDir </> dbSchemaChecksumFileInProjectRootDir
+
+checkIfDbSchemaMatchesChecksum' :: Maybe FilePath -> Maybe FilePath -> ExceptT String IO ()
+checkIfDbSchemaMatchesChecksum' (Just dbSchemaFp) (Just dbSchemaChecksumFp) = do
+  dbSchemaFileChecksum <- liftIO $ calculateChecksumStringForFile dbSchemaFp
+  dbChecksumFileContents <- liftIO $ readFile dbSchemaChecksumFp
+  if dbSchemaFileChecksum == dbChecksumFileContents
+    then return ()
+    else throwError "Your Prisma schema has changed, you should run wasp db migrate-dev."
+checkIfDbSchemaMatchesChecksum' (Just _dbSchemaFp) Nothing =
+  -- Cannot throw the error below because they may have no entities in the schema.prisma file.
```

Comment:
The problem here is both: (1) legacy projects won't have a checksum (but that is minor), but more importantly (2) a brand new project that has no entities will still have a schema, but no checksum (since they don't need to, and indeed cannot migrate). We have no way to know in that case, unless we parse the file, if it "should" have to be migrated (and thus have a checksum) or not without looking at the contents of the schema.prisma. Not sure if we want to go down that path in case that file changes in the future? Feels brittle. So here, we just currently don't issue any error.

I plan to make a state diagram today to validate what is happening right now vs all the permutations, since as you noted with the checksum in the repo, it gets complicated fast. 😅

Reply:
Can empty schema.prisma file exist? I would expect that in such case, there is no schema.prisma file? But now that I think about it, I have never checked what is the situation in such case.

Reply:
There should always be a schema.prisma file with some contents, even if they have no entities. Here is what an empty one looks like after a `wasp new` and `wasp start`:

```
datasource db {
  provider = "sqlite"
  url      = "file:./dev.db"
}

generator client {
  provider = "prisma-client-js"
  output = "../server/node_modules/.prisma/client"
}
```

Reply:
Ha interesting I didn't know this!
In such case, should migration be done or not? What happens if `migrate` is called? Migration that removes everything from the db? Hmmm I am not sure how to think about it.

Reply:
Ha, yeah this was a fun wrinkle! If we call `db mgirate-dev` in this state (with no entities) we get a Prisma error:

```
Performing migration...
Prisma schema loaded from ../db/schema.prisma
Datasource "db": SQLite database "dev.db" at "file:./dev.db"

SQLite database dev.db created at file:./dev.db

Already in sync, no schema change or pending migration was found.

Running generate... (Use --skip-generate to skip the generators)
Error:
You don't have any models defined in your schema.prisma, so nothing will be generated.
You can define a model like this:

model User {
  id    Int     @id @default(autoincrement())
  email String  @unique
  name  String?
}

More information in our documentation:
https://pris.ly/d/prisma-schema

Error: Migrate dev failed: Migrate (dev) failed with exit code: 1
```

So in this case, they won't have a checksum file (since migrate cannot succeed), and we should _not_ warn them to migrate since they cannot. Here we can do the "check for entities" approach of looking for a migrations dir (or using AppSpec if we had access in a Warnings world).

Reply:
Here is a comment above the actual check that covers in detail some of these assumptions and the behavior we have now:

```haskell
-- | This function warns the user to run `wasp db migrate-dev` under certain conditions. But first, a few preliminaries:
--   - This function relies on the latest schema.prisma, and thus should run after the writing of FileDrafts has occurred.
--   - A non-empty schema.prisma should always exist in the generated project dir, even if the user has no entities defined.
--     `wasp new` will create one with a `datasource db {}` and `generator client {}`, for example.
--   - A schema.prisma.wasp-checksum file is created/updated each time `wasp db migrate-dev` is run locally and contains checksum(schema.prisma).
--
-- Given that, there are two cases in which we wish to warn the user to run `wasp db migrate-dev`:
-- (1) If schema.prisma.wasp-checksum exists, but is not equal to checksum(schema.prisma), we know they made changes to schema.prisma and should migrate.
-- (2) If schema.prisma.wasp-checksum does not exist, but the user has entities defined in schema.prisma (and thus, AppSpec).
--     This could imply they have never migrated locally, or that they have but are simply missing their generated project dir.
--     Common scenarios for the second warning include:
--       - After a fresh checkout, or after `wasp clean`; possible false positives in these cases, but for safety, it's still preferable to warn.
--       - When they previously had no entities and just added their first.
checkIfDbSchemaWasUsedForLastMigration :: AppSpec -> Path' Abs (Dir ProjectRootDir) -> IO [GeneratorWarning]
```